File: back-end/src/main/java/nl/hva/backend/domain/models/game/action/impl/money/moneyExchangeImpl/PayTaxAction.java
package nl.hva.backend.domain.models.game.action.impl.money.moneyExchangeImpl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.Player;
import nl.hva.backend.domain.models.game.action.impl.money.MoneyExchangeAction;

import javax.persistence.Entity;

/**
 * A representation, of the action, of a tax payment within a GameSession,
 * this, for example occurs when landing on TaxLocation on the GameBoard.
 *
 * @author Hamza el Haouti
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
public class PayTaxAction extends MoneyExchangeAction {

    /**
     * Constructor for the manual creation of an instance, that is intended to be persisted to JPA.
     *
     * @param payee  The party paying, a Player instance.
     * @param amount The amount, to be subtracted from the payee, and given to the bank.
     */
    public PayTaxAction(
            Player payee,
            double amount
    ) {
        super(payee, null, amount);
    }

}

File: front-end/src/app/domain/models/game/action/impl/money/PayTaxAction.ts
import {Player} from "../../../Player";
import {MoneyExchangeAction} from "./MoneyExchangeAction";

/**
 * A representation, of the action, of a tax payment within a GameSession,
 * this, for example occurs when landing on TaxLocation on the GameBoard.
 *
 * @author Hamza el Haouti
 */
export class PayTaxAction extends MoneyExchangeAction {

  constructor(
    id: number,
    payee: Player,
    performedAt: Date,
    amount: number
  ) {
    super(PayTaxAction.name, id, payee, performedAt, null, amount);
  }

  public static trueCopy(action: PayTaxAction): PayTaxAction | null {
    return new PayTaxAction(
      action.id,
      Player.trueCopy(action.actor),
      new Date(action.performedAt),
      action.amount
    );
  }

  public override toString(): string {
    return `${this.actor.user.username} payed a tax of $${this.amount}.`;
  }

}
